// @solana-mobile/mobile-wallet-adapter-protocol-web3js/lib/cjs/index.native.js
'use strict';

var web3_js = require('@solana/web3.js');
var mobileWalletAdapterProtocol = require('@solana-mobile/mobile-wallet-adapter-protocol');
var bs58 = require('bs58');
var jsBase64 = require('js-base64');

function getPayloadFromTransaction(transaction) {
    const serializedTransaction = 'version' in transaction
        ? transaction.serialize()
        : transaction.serialize({
            requireAllSignatures: false,
            verifySignatures: false,
        });
    const payload = jsBase64.fromUint8Array(serializedTransaction);
    return payload;
}
function getTransactionFromWireMessage(byteArray) {
    const numSignatures = byteArray[0];
    const messageOffset = numSignatures * web3_js.SIGNATURE_LENGTH_IN_BYTES + 1;
    const version = web3_js.VersionedMessage.deserializeMessageVersion(byteArray.slice(messageOffset, byteArray.length));
    if (version === 'legacy') {
        return web3_js.Transaction.from(byteArray);
    }
    else {
        return web3_js.VersionedTransaction.deserialize(byteArray);
    }
}
async function transact(callback, config) {
    const augmentedCallback = (wallet) => {
        return callback(augmentWalletAPI(wallet));
    };
    return await mobileWalletAdapterProtocol.transact(augmentedCallback, config);
}
async function startRemoteScenario(config) {
    const { wallet, close, associationUrl } = await mobileWalletAdapterProtocol.startRemoteScenario(config);
    const augmentedPromise = wallet.then((wallet) => {
        return augmentWalletAPI(wallet);
    });
    return { wallet: augmentedPromise, close, associationUrl };
}
function augmentWalletAPI(wallet) {
    return new Proxy({}, {
        get(target, p) {
            if (target[p] == null) {
                switch (p) {
                    case 'signAndSendTransactions':
                        target[p] = async function ({ minContextSlot, commitment, skipPreflight, maxRetries, waitForCommitmentToSendNextTransaction, transactions, ...rest }) {
                            const payloads = transactions.map(getPayloadFromTransaction);
                            const options = {
                                min_context_slot: minContextSlot,
                                commitment: commitment,
                                skip_preflight: skipPreflight,
                                max_retries: maxRetries,
                                wait_for_commitment_to_send_next_transaction: waitForCommitmentToSendNextTransaction
                            };
                            const { signatures: base64EncodedSignatures } = await wallet.signAndSendTransactions({
                                ...rest,
                                ...(Object.values(options).some(element => element != null)
                                    ? { options: options }
                                    : null),
                                payloads,
                            });
                            const signatures = base64EncodedSignatures.map(jsBase64.toUint8Array).map(bs58.encode);
                            return signatures;
                        };
                        break;
                    case 'signMessages':
                        target[p] = async function ({ payloads, ...rest }) {
                            const base64EncodedPayloads = payloads.map(jsBase64.fromUint8Array);
                            const { signed_payloads: base64EncodedSignedMessages } = await wallet.signMessages({
                                ...rest,
                                payloads: base64EncodedPayloads,
                            });
                            const signedMessages = base64EncodedSignedMessages.map(jsBase64.toUint8Array);
                            return signedMessages;
                        };
                        break;
                    case 'signTransactions':
                        target[p] = async function ({ transactions, ...rest }) {
                            const payloads = transactions.map(getPayloadFromTransaction);
                            const { signed_payloads: base64EncodedCompiledTransactions } = await wallet.signTransactions({
                                ...rest,
                                payloads,
                            });
                            const compiledTransactions = base64EncodedCompiledTransactions.map(jsBase64.toUint8Array);
                            const signedTransactions = compiledTransactions.map(getTransactionFromWireMessage);
                            return signedTransactions;
                        };
                        break;
                    default: {
                        target[p] = wallet[p];
                        break;
                    }
                }
            }
            return target[p];
        },
        defineProperty() {
            return false;
        },
        deleteProperty() {
            return false;
        },
    });
}

exports.startRemoteScenario = startRemoteScenario;
exports.transact = transact;
